typed-racket-lib/typed-racket/rep/type-rep.rkt

```diff
@@ -1077,92 +1077,141 @@
        -Bottom
        (make-Distinction nm id ty))])
 
-;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;;************************************************************
+;; Locally Nameles Type Variable Abstraction/Instantiation
+;;************************************************************
+
+
+;; abstract
+;;
+;; asbtracts type variable 'name'
+;; to De Bruijn index 0 in 'initial'
+(define/cond-contract (abstract name initial)
+  (-> symbol? Rep? Rep?)
+  (abstract-many (list name) initial))
+
+;; abstract-many
+;;
+;; abstracts the type variable names from 'names-to-abstract'
+;; to de bruijn indices in 'initial'.
+;; Specifically, if n = (length names-to-abstract) then
+;; names-to-abstract[0] gets mapped to n-1
+;; names-to-abstract[1] gets mapped to n-2
+;; ...
+;; names-to-abstract[n-1] gets mapped to 0
+(define/cond-contract (abstract-many names-to-abstract initial)
+  (-> (listof symbol?) Rep? Rep?)
+  (define n-1 (sub1 (length names-to-abstract)))
+  (define (abstract-name name lvl default dotted?)
+    (match (index-of names-to-abstract name eq?)
```

Comment:
IMHO it wouldn't actually simplify anything, it would just move around the logic locally (and it allocates more)